File: data_mcp/gpt-5/go/rep2/main.go
package main

import (
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"errors"
	"flag"
	"fmt"
	"io"
	"log"
	"net/http"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"

	"golang.org/x/crypto/bcrypt"
)

type User struct {
	ID       int    `json:"id"`
	Username string `json:"username"`
}

type userRecord struct {
	User
	PasswordHash []byte
}

type Todo struct {
	ID          int    `json:"id"`
	Title       string `json:"title"`
	Description string `json:"description"`
	Completed   bool   `json:"completed"`
	CreatedAt   string `json:"created_at"`
	UpdatedAt   string `json:"updated_at"`
}

type server struct {
	mu sync.RWMutex

	// users by id
	users map[int]*userRecord
	// username -> id
	usernameIndex map[string]int
	// next user id
	nextUserID int

	// todos by id
	todos map[int]*Todo
	// todo ownership: todo id -> user id
	todoOwner map[int]int
	nextTodoID int

	// sessions token -> user id
	sessions map[string]int
}

func newServer() *server {
	return &server{
		users:         make(map[int]*userRecord),
		usernameIndex: make(map[string]int),
		nextUserID:    1,
		todos:         make(map[int]*Todo),
		todoOwner:     make(map[int]int),
		nextTodoID:    1,
		sessions:      make(map[string]int),
	}
}

func (s *server) nowTS() string {
	// UTC with second precision
	return time.Now().UTC().Format("2006-01-02T15:04:05Z")
}

func writeJSON(w http.ResponseWriter, status int, v any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	if v == nil {
		return
	}
	enc := json.NewEncoder(w)
	enc.SetEscapeHTML(false)
	_ = enc.Encode(v)
}

func writeError(w http.ResponseWriter, status int, msg string) {
	writeJSON(w, status, map[string]string{"error": msg})
}

func parseJSON(r *http.Request, dst any) error {
	defer r.Body.Close()
	dec := json.NewDecoder(r.Body)
	dec.DisallowUnknownFields()
	if err := dec.Decode(dst); err != nil {
		return err
	}
	// ensure no trailing data
	if _, err := dec.Token(); err != io.EOF {
		if err == nil {
			return errors.New("invalid JSON: extra data")
		}
		return err
	}
	return nil
}

func (s *server) generateToken() (string, error) {
	b := make([]byte, 16) // 128-bit
	if _, err := rand.Read(b); err != nil {
		return "", err
	}
	return hex.EncodeToString(b), nil
}

func (s *server) getAuthUserID(r *http.Request) (int, bool) {
	cookie, err := r.Cookie("session_id")
	if err != nil || cookie.Value == "" {
		return 0, false
	}
	s.mu.RLock()
	defer s.mu.RUnlock()
	uid, ok := s.sessions[cookie.Value]
	return uid, ok
}

func (s *server) requireAuth(next http.HandlerFunc) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		if uid, ok := s.getAuthUserID(r); ok {
			// attach uid to context using request headers map to avoid context types
			r = r.Clone(r.Context())
			r.Header.Set("X-Auth-UserID", strconv.Itoa(uid))
			next(w, r)
			return
		}
		writeError(w, http.StatusUnauthorized, "Authentication required")
	}
}

func (s *server) handleRegister(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		writeError(w, http.StatusNotFound, "Not found")
		return
	}
	var req struct {
		Username string `json:"username"`
		Password string `json:"password"`
	}
	if err := parseJSON(r, &req); err != nil {
		writeError(w, http.StatusBadRequest, "Invalid JSON")
		return
	}
	uname := strings.TrimSpace(req.Username)
	if !regexp.MustCompile(`^[a-zA-Z0-9_]{3,50}$`).MatchString(uname) {
		writeError(w, http.StatusBadRequest, "Invalid username")
		return
	}
	if len(req.Password) < 8 {
		writeError(w, http.StatusBadRequest, "Password too short")
		return
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	if _, exists := s.usernameIndex[uname]; exists {
		writeError(w, http.StatusConflict, "Username already exists")
		return
	}
	hash, err := bcrypt.GenerateFromPassword([]byte(req.Password), bcrypt.DefaultCost)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "Internal error")
		return
	}
	id := s.nextUserID
	s.nextUserID++
	rec := &userRecord{User: User{ID: id, Username: uname}, PasswordHash: hash}
	s.users[id] = rec
	s.usernameIndex[uname] = id
	writeJSON(w, http.StatusCreated, rec.User)
}

func (s *server) handleLogin(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		writeError(w, http.StatusNotFound, "Not found")
		return
	}
	var req struct {
		Username string `json:"username"`
		Password string `json:"password"`
	}
	if err := parseJSON(r, &req); err != nil {
		writeError(w, http.StatusBadRequest, "Invalid JSON")
		return
	}
	s.mu.RLock()
	uid, ok := s.usernameIndex[req.Username]
	var rec *userRecord
	if ok {
		rec = s.users[uid]
	}
	s.mu.RUnlock()
	if !ok || bcrypt.CompareHashAndPassword(rec.PasswordHash, []byte(req.Password)) != nil {
		writeError(w, http.StatusUnauthorized, "Invalid credentials")
		return
	}
	token, err := s.generateToken()
	if err != nil {
		writeError(w, http.StatusInternalServerError, "Internal error")
		return
	}
	s.mu.Lock()
	s.sessions[token] = uid
	s.mu.Unlock()
	http.SetCookie(w, &http.Cookie{Name: "session_id", Value: token, Path: "/", HttpOnly: true})
	writeJSON(w, http.StatusOK, rec.User)
}

func (s *server) handleLogout(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		writeError(w, http.StatusNotFound, "Not found")
		return
	}
	cookie, err := r.Cookie("session_id")
	if err != nil {
		writeError(w, http.StatusUnauthorized, "Authentication required")
		return
	}
	s.mu.Lock()
	delete(s.sessions, cookie.Value)
	s.mu.Unlock()
	writeJSON(w, http.StatusOK, map[string]any{})
}

func (s *server) currentUser(r *http.Request) (User, bool) {
	uidStr := r.Header.Get("X-Auth-UserID")
	uid, _ := strconv.Atoi(uidStr)
	if uid == 0 {
		return User{}, false
	}
	s.mu.RLock()
	rec, ok := s.users[uid]
	s.mu.RUnlock()
	if !ok {
		return User{}, false
	}
	return rec.User, true
}

func (s *server) handleMe(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		writeError(w, http.StatusNotFound, "Not found")
		return
	}
	user, ok := s.currentUser(r)
	if !ok {
		writeError(w, http.StatusUnauthorized, "Authentication required")
		return
	}
	writeJSON(w, http.StatusOK, user)
}

func (s *server) handlePassword(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPut {
		writeError(w, http.StatusNotFound, "Not found")
		return
	}
	user, ok := s.currentUser(r)
	if !ok {
		writeError(w, http.StatusUnauthorized, "Authentication required")
		return
	}
	var req struct {
		Old string `json:"old_password"`
		New string `json:"new_password"`
	}
	if err := parseJSON(r, &req); err != nil {
		writeError(w, http.StatusBadRequest, "Invalid JSON")
		return
	}
	s.mu.RLock()
	rec := s.users[user.ID]
	s.mu.RUnlock()
	if bcrypt.CompareHashAndPassword(rec.PasswordHash, []byte(req.Old)) != nil {
		writeError(w, http.StatusUnauthorized, "Invalid credentials")
		return
	}
	if len(req.New) < 8 {
		writeError(w, http.StatusBadRequest, "Password too short")
		return
	}
	hash, err := bcrypt.GenerateFromPassword([]byte(req.New), bcrypt.DefaultCost)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "Internal error")
		return
	}
	s.mu.Lock()
	rec.PasswordHash = hash
	s.mu.Unlock()
	writeJSON(w, http.StatusOK, map[string]any{})
}

func (s *server) handleTodos(w http.ResponseWriter, r *http.Request) {
	user, ok := s.currentUser(r)
	if !ok {
		writeError(w, http.StatusUnauthorized, "Authentication required")
		return
	}
	switch r.Method {
	case http.MethodGet:
		// list
		s.mu.RLock()
		list := make([]*Todo, 0)
		for id, t := range s.todos {
			if s.todoOwner[id] == user.ID {
				copy := *t
				list = append(list, &copy)
			}
		}
		s.mu.RUnlock()
		sort.Slice(list, func(i, j int) bool { return list[i].ID < list[j].ID })
		writeJSON(w, http.StatusOK, list)
	case http.MethodPost:
		var req struct {
			Title       string `json:"title"`
			Description string `json:"description"`
		}
		if err := parseJSON(r, &req); err != nil {
			writeError(w, http.StatusBadRequest, "Invalid JSON")
			return
		}
		title := strings.TrimSpace(req.Title)
		if title == "" {
			writeError(w, http.StatusBadRequest, "Title is required")
			return
		}
		desc := req.Description
		if desc == "" {
			desc = ""
		}
		now := s.nowTS()
		s.mu.Lock()
		id := s.nextTodoID
		s.nextTodoID++
		t := &Todo{ID: id, Title: title, Description: desc, Completed: false, CreatedAt: now, UpdatedAt: now}
		s.todos[id] = t
		s.todoOwner[id] = user.ID
		s.mu.Unlock()
		writeJSON(w, http.StatusCreated, t)
	default:
		writeError(w, http.StatusNotFound, "Not found")
	}
}

func parseTodoID(path string) (int, bool) {
	// expect /todos/:id exactly
	parts := strings.Split(strings.Trim(path, "/"), "/")
	if len(parts) != 2 || parts[0] != "todos" {
		return 0, false
	}
	id, err := strconv.Atoi(parts[1])
	if err != nil || id <= 0 {
		return 0, false
	}
	return id, true
}

func (s *server) handleTodoByID(w http.ResponseWriter, r *http.Request) {
	user, ok := s.currentUser(r)
	if !ok {
		writeError(w, http.StatusUnauthorized, "Authentication required")
		return
	}
	id, ok := parseTodoID(r.URL.Path)
	if !ok {
		writeError(w, http.StatusNotFound, "Not found")
		return
	}
	s.mu.RLock()
	t, exists := s.todos[id]
	owner := s.todoOwner[id]
	s.mu.RUnlock()
	if !exists || owner != user.ID {
		writeError(w, http.StatusNotFound, "Todo not found")
		return
	}
	switch r.Method {
	case http.MethodGet:
		writeJSON(w, http.StatusOK, t)
	case http.MethodPut:
		var req struct {
			Title       *string `json:"title"`
			Description *string `json:"description"`
			Completed   *bool   `json:"completed"`
		}
		if err := parseJSON(r, &req); err != nil {
			writeError(w, http.StatusBadRequest, "Invalid JSON")
			return
		}
		s.mu.Lock()
		defer s.mu.Unlock()
		// re-fetch to ensure current and owner check again
		t2, exists2 := s.todos[id]
		if !exists2 || s.todoOwner[id] != user.ID {
			writeError(w, http.StatusNotFound, "Todo not found")
			return
		}
		if req.Title != nil {
			title := strings.TrimSpace(*req.Title)
			if title == "" {
				writeError(w, http.StatusBadRequest, "Title is required")
				return
			}
			t2.Title = title
		}
		if req.Description != nil {
			t2.Description = *req.Description
		}
		if req.Completed != nil {
			t2.Completed = *req.Completed
		}
		t2.UpdatedAt = s.nowTS()
		writeJSON(w, http.StatusOK, t2)
	case http.MethodDelete:
		s.mu.Lock()
		defer s.mu.Unlock()
		// verify still exists and owner
		if _, ok := s.todos[id]; !ok || s.todoOwner[id] != user.ID {
			writeError(w, http.StatusNotFound, "Todo not found")
			return
		}
		delete(s.todos, id)
		delete(s.todoOwner, id)
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusNoContent)
	default:
		writeError(w, http.StatusNotFound, "Not found")
	}
}

func (s *server) routes() http.Handler {
	mux := http.NewServeMux()
	mux.HandleFunc("/register", s.handleRegister)
	mux.HandleFunc("/login", s.handleLogin)
	mux.HandleFunc("/logout", s.requireAuth(s.handleLogout))
	mux.HandleFunc("/me", s.requireAuth(s.handleMe))
	mux.HandleFunc("/password", s.requireAuth(s.handlePassword))
	mux.HandleFunc("/todos", s.requireAuth(s.handleTodos))
	mux.HandleFunc("/todos/", s.requireAuth(s.handleTodoByID))
	// any other path
	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		writeError(w, http.StatusNotFound, "Not found")
	})
	return contentTypeJSONMiddleware(mux)
}

func contentTypeJSONMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		// Ensure all non-DELETE responses have application/json
		rec := &statusRecorder{ResponseWriter: w, status: http.StatusOK}
		next.ServeHTTP(rec, r)
		if r.Method != http.MethodDelete {
			if ct := w.Header().Get("Content-Type"); ct == "" {
				w.Header().Set("Content-Type", "application/json")
			}
		}
	})
}

type statusRecorder struct {
	http.ResponseWriter
	status int
}

func (s *statusRecorder) WriteHeader(code int) {
	s.status = code
	s.ResponseWriter.WriteHeader(code)
}

func main() {
	port := flag.Int("port", 8080, "port to listen on")
	flag.Parse()

	s := newServer()

	addr := fmt.Sprintf("0.0.0.0:%d", *port)
	server := &http.Server{
		Addr:              addr,
		Handler:           s.routes(),
		ReadHeaderTimeout: 5 * time.Second,
	}
	log.Printf("Listening on %s", addr)
	if err := server.ListenAndServe(); err != nil && !errors.Is(err, http.ErrServerClosed) {
		log.Fatalf("server error: %v", err)
	}
}

File: data_mcp/gpt-5/java/rep2/src/Main.java
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;
import com.google.gson.*;

import java.io.*;
import java.net.*;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.concurrent.*;
import java.util.regex.Pattern;

public class Main {
    // Data models
    static class User {
        final int id;
        final String username;
        String passwordHash; // stored as plain for this exercise
        User(int id, String username, String passwordHash) {
            this.id = id; this.username = username; this.passwordHash = passwordHash;
        }
    }
    static class Todo {
        final int id;
        final int userId;
        String title;
        String description;
        boolean completed;
        String createdAt;
        String updatedAt;
        Todo(int id, int userId, String title, String description, boolean completed, String ts) {
            this.id=id; this.userId=userId; this.title=title; this.description=description; this.completed=completed; this.createdAt=ts; this.updatedAt=ts;
        }
        JsonObject toJson() {
            JsonObject o = new JsonObject();
            o.addProperty("id", id);
            o.addProperty("title", title);
            o.addProperty("description", description);
            o.addProperty("completed", completed);
            o.addProperty("created_at", createdAt);
            o.addProperty("updated_at", updatedAt);
            return o;
        }
    }

    static class State {
        final ConcurrentMap<String, Integer> sessions = new ConcurrentHashMap<>(); // token -> userId
        final ConcurrentMap<String, User> usersByName = new ConcurrentHashMap<>();
        final ConcurrentMap<Integer, User> usersById = new ConcurrentHashMap<>();
        final ConcurrentMap<Integer, Todo> todosById = new ConcurrentHashMap<>();
        final java.util.concurrent.atomic.AtomicInteger userSeq = new java.util.concurrent.atomic.AtomicInteger(0);
        final java.util.concurrent.atomic.AtomicInteger todoSeq = new java.util.concurrent.atomic.AtomicInteger(0);
    }

    static final Gson gson = new GsonBuilder().serializeNulls().create();
    static final Pattern USERNAME_RE = Pattern.compile("^[a-zA-Z0-9_]{3,50}$");
    static final DateTimeFormatter TS_FMT = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'").withZone(ZoneOffset.UTC);

    public static void main(String[] args) throws Exception {
        int port = 8080;
        for (int i=0;i<args.length;i++) {
            if ("--port".equals(args[i]) && i+1<args.length) {
                port = Integer.parseInt(args[i+1]);
            }
        }
        InetSocketAddress addr = new InetSocketAddress("0.0.0.0", port);
        HttpServer server = HttpServer.create(addr, 0);
        State state = new State();
        server.createContext("/register", exchange -> handleRegister(exchange, state));
        server.createContext("/login", exchange -> handleLogin(exchange, state));
        server.createContext("/logout", exchange -> withAuth(exchange, state, (ex, user)-> handleLogout(ex, state, user)));
        server.createContext("/me", exchange -> withAuth(exchange, state, (ex, user)-> handleMe(ex, user)));
        server.createContext("/password", exchange -> withAuth(exchange, state, (ex, user)-> handlePassword(ex, state, user)));
        server.createContext("/todos", exchange -> withAuth(exchange, state, (ex, user)-> handleTodos(ex, state, user)));
        server.createContext("/todos/", exchange -> withAuth(exchange, state, (ex, user)-> handleTodoById(ex, state, user)));
        server.createContext("/", exchange -> {
            // Unknown endpoint
            sendJson(exchange, 404, error("Not found"));
        });
        server.setExecutor(Executors.newCachedThreadPool());
        server.start();
        System.out.println("Server started on 0.0.0.0:"+port);
    }

    interface AuthedHandler { void handle(HttpExchange ex, User user) throws IOException; }

    static void withAuth(HttpExchange ex, State state, AuthedHandler next) throws IOException {
        Optional<String> sidOpt = readSessionCookie(ex);
        if (!sidOpt.isPresent()) {
            sendAuthRequired(ex); return;
        }
        Integer uid = state.sessions.get(sidOpt.get());
        if (uid == null) { sendAuthRequired(ex); return; }
        User user = state.usersById.get(uid);
        if (user == null) { sendAuthRequired(ex); return; }
        next.handle(ex, user);
    }

    static void handleRegister(HttpExchange ex, State state) throws IOException {
        if (!"POST".equals(ex.getRequestMethod())) { sendJson(ex, 405, error("Method not allowed")); return; }
        JsonObject body = readJson(ex);
        if (body == null) return; // error already sent
        String username = optString(body, "username");
        String password = optString(body, "password");
        if (username == null || !USERNAME_RE.matcher(username).matches()) {
            sendJson(ex, 400, error("Invalid username")); return;
        }
        if (password == null || password.length() < 8) {
            sendJson(ex, 400, error("Password too short")); return;
        }
        // enforce unique
        synchronized (state.usersByName) {
            if (state.usersByName.containsKey(username)) {
                sendJson(ex, 409, error("Username already exists")); return;
            }
            int id = state.userSeq.incrementAndGet();
            User u = new User(id, username, password);
            state.usersByName.put(username, u);
            state.usersById.put(id, u);
            JsonObject resp = new JsonObject();
            resp.addProperty("id", id);
            resp.addProperty("username", username);
            sendJson(ex, 201, resp);
        }
    }

    static void handleLogin(HttpExchange ex, State state) throws IOException {
        if (!"POST".equals(ex.getRequestMethod())) { sendJson(ex, 405, error("Method not allowed")); return; }
        JsonObject body = readJson(ex);
        if (body == null) return;
        String username = optString(body, "username");
        String password = optString(body, "password");
        if (username == null || password == null) { sendJson(ex, 401, error("Invalid credentials")); return; }
        User u = state.usersByName.get(username);
        if (u == null || !Objects.equals(u.passwordHash, password)) {
            sendJsonWithSetCookie(ex, 401, error("Invalid credentials"), null); return;
        }
        String token = UUID.randomUUID().toString().replace("-", "");
        state.sessions.put(token, u.id);
        JsonObject resp = new JsonObject();
        resp.addProperty("id", u.id);
        resp.addProperty("username", u.username);
        String setCookie = "session_id="+token+"; Path=/; HttpOnly";
        sendJsonWithSetCookie(ex, 200, resp, setCookie);
    }

    static void handleLogout(HttpExchange ex, State state, User user) throws IOException {
        if (!"POST".equals(ex.getRequestMethod())) { sendJson(ex, 405, error("Method not allowed")); return; }
        Optional<String> sid = readSessionCookie(ex);
        sid.ifPresent(token -> state.sessions.remove(token));
        sendJson(ex, 200, new JsonObject());
    }

    static void handleMe(HttpExchange ex, User user) throws IOException {
        if (!"GET".equals(ex.getRequestMethod())) { sendJson(ex, 405, error("Method not allowed")); return; }
        JsonObject resp = new JsonObject();
        resp.addProperty("id", user.id);
        resp.addProperty("username", user.username);
        sendJson(ex, 200, resp);
    }

    static void handlePassword(HttpExchange ex, State state, User user) throws IOException {
        if (!"PUT".equals(ex.getRequestMethod())) { sendJson(ex, 405, error("Method not allowed")); return; }
        JsonObject body = readJson(ex);
        if (body == null) return;
        String oldp = optString(body, "old_password");
        String newp = optString(body, "new_password");
        if (oldp == null || !Objects.equals(oldp, user.passwordHash)) { sendJson(ex, 401, error("Invalid credentials")); return; }
        if (newp == null || newp.length() < 8) { sendJson(ex, 400, error("Password too short")); return; }
        user.passwordHash = newp;
        sendJson(ex, 200, new JsonObject());
    }

    static void handleTodos(HttpExchange ex, State state, User user) throws IOException {
        String method = ex.getRequestMethod();
        if ("GET".equals(method)) {
            List<Todo> list = new ArrayList<>();
            for (Todo t : state.todosById.values()) if (t.userId == user.id) list.add(t);
            list.sort(Comparator.comparingInt(t -> t.id));
            JsonArray arr = new JsonArray();
            for (Todo t : list) arr.add(t.toJson());
            sendJson(ex, 200, arr);
            return;
        } else if ("POST".equals(method)) {
            JsonObject body = readJson(ex);
            if (body == null) return;
            String title = optString(body, "title");
            if (title == null || title.trim().isEmpty()) { sendJson(ex, 400, error("Title is required")); return; }
            String description = optString(body, "description");
            if (description == null) description = "";
            int id = state.todoSeq.incrementAndGet();
            String ts = TS_FMT.format(Instant.now());
            Todo t = new Todo(id, user.id, title, description, false, ts);
            state.todosById.put(id, t);
            sendJson(ex, 201, t.toJson());
            return;
        } else {
            sendJson(ex, 405, error("Method not allowed")); return;
        }
    }

    static void handleTodoById(HttpExchange ex, State state, User user) throws IOException {
        String path = ex.getRequestURI().getPath();
        // path like /todos/123
        String[] parts = path.split("/");
        if (parts.length < 3) { sendJson(ex, 404, error("Not found")); return; }
        int id;
        try { id = Integer.parseInt(parts[2]); } catch (Exception e) { sendJson(ex, 404, error("Todo not found")); return; }
        Todo t = state.todosById.get(id);
        if (t == null || t.userId != user.id) { sendJson(ex, 404, error("Todo not found")); return; }
        String method = ex.getRequestMethod();
        switch (method) {
            case "GET":
                sendJson(ex, 200, t.toJson());
                return;
            case "PUT":
                JsonObject body = readJson(ex);
                if (body == null) return;
                if (body.has("title")) {
                    String title = optString(body, "title");
                    if (title == null || title.trim().isEmpty()) { sendJson(ex, 400, error("Title is required")); return; }
                    t.title = title;
                }
                if (body.has("description")) {
                    String description = optString(body, "description");
                    t.description = description == null ? "" : description;
                }
                if (body.has("completed")) {
                    JsonElement c = body.get("completed");
                    if (c != null && !c.isJsonNull() && c.isJsonPrimitive() && ((JsonPrimitive)c).isBoolean()) {
                        t.completed = c.getAsBoolean();
                    }
                }
                t.updatedAt = TS_FMT.format(Instant.now());
                sendJson(ex, 200, t.toJson());
                return;
            case "DELETE":
                state.todosById.remove(id);
                sendNoContent(ex);
                return;
            default:
                sendJson(ex, 405, error("Method not allowed"));
        }
    }

    // Utilities
    static JsonObject readJson(HttpExchange ex) throws IOException {
        try (InputStream is = ex.getRequestBody()) {
            String body = new String(is.readAllBytes(), StandardCharsets.UTF_8);
            if (body == null || body.isEmpty()) return new JsonObject();
            try {
                JsonElement el = JsonParser.parseString(body);
                if (!el.isJsonObject()) { sendJson(ex, 400, error("Invalid JSON")); return null; }
                return el.getAsJsonObject();
            } catch (JsonSyntaxException e) {
                sendJson(ex, 400, error("Invalid JSON")); return null;
            }
        }
    }

    static String optString(JsonObject o, String key) {
        if (o == null || !o.has(key)) return null;
        JsonElement v = o.get(key);
        if (v == null || v.isJsonNull()) return null;
        if (v.isJsonPrimitive()) return v.getAsString();
        return null;
    }

    static JsonObject error(String msg) {
        JsonObject o = new JsonObject();
        o.addProperty("error", msg);
        return o;
    }

    static Optional<String> readSessionCookie(HttpExchange ex) {
        List<String> cookies = ex.getRequestHeaders().get("Cookie");
        if (cookies == null) return Optional.empty();
        for (String c : cookies) {
            String[] parts = c.split(";\\s*");
            for (String p : parts) {
                int eq = p.indexOf('=');
                if (eq <= 0) continue;
                String name = p.substring(0, eq).trim();
                String val = p.substring(eq+1).trim();
                if ("session_id".equals(name)) return Optional.of(val);
            }
        }
        return Optional.empty();
    }

    static void sendAuthRequired(HttpExchange ex) throws IOException {
        sendJson(ex, 401, error("Authentication required"));
    }

    static void sendJson(HttpExchange ex, int status, JsonElement body) throws IOException {
        byte[] data = gson.toJson(body).getBytes(StandardCharsets.UTF_8);
        ex.getResponseHeaders().set("Content-Type", "application/json");
        ex.sendResponseHeaders(status, data.length);
        try (OutputStream os = ex.getResponseBody()) { os.write(data); }
    }

    static void sendJsonWithSetCookie(HttpExchange ex, int status, JsonElement body, String setCookie) throws IOException {
        byte[] data = gson.toJson(body).getBytes(StandardCharsets.UTF_8);
        ex.getResponseHeaders().set("Content-Type", "application/json");
        if (setCookie != null) ex.getResponseHeaders().add("Set-Cookie", setCookie);
        ex.sendResponseHeaders(status, data.length);
        try (OutputStream os = ex.getResponseBody()) { os.write(data); }
    }

    static void sendNoContent(HttpExchange ex) throws IOException {
        ex.getResponseHeaders().set("Content-Type", "application/json");
        ex.sendResponseHeaders(204, -1); // no body
        ex.close();
    }
}
